nixos/lib/testing/interactive.nix

```diff
@@ -1,6 +1,97 @@
 { config, lib, moduleType, hostPkgs, ... }:
 let
-  inherit (lib) mkOption types mdDoc;
+  inherit (lib) mkOption mkDefault types mdDoc concatStringsSep mapAttrsToList;
+  inherit (builtins) toString attrNames;
+  finalNodes = config.interactive.driver.nodes;
+
+  # allow accessing machines through a jumphost whose ssh port is forwarded to from the host
+  sshConfig = hostPkgs.writeText "ssh-config" ''
+    Host *
+      User root
+      StrictHostKeyChecking no
+      BatchMode yes
+      ConnectTimeout 20
+      UserKnownHostsFile=/dev/null
+      LogLevel Error # no "added to known hosts"
+    Host redeploy_jumphost
+      HostName localhost
+      Port ${
+        let
+          fwds = config.interactive.nodes.redeploy_jumphost.virtualisation.forwardPorts;
+          sshFwds = builtins.filter (fwd: fwd.guest.port == 22) fwds;
+          sshFwd = builtins.head sshFwds;
+        in
+          toString sshFwd.host.port
+        }
+    Host * !redeploy_jumphost
+      ProxyJump redeploy_jumphost
+  '';
+
+  redeploy = hostPkgs.writeShellScriptBin "redeploy" ''
+    # create an association array from machine names to the path to their
+    # configuration in the nix store
+    declare -A configPaths=(${
+      concatStringsSep " "
+        (mapAttrsToList
+          (n: v: ''["${n}"]="${v.system.build.toplevel}"'')
+          finalNodes)
+    })
+
+    redeploy_one() {
+      machine="$1"
+      echo "pushing new config to '$machine'"
+
+      if [ -z ''${configPaths[$machine]+x} ]; then
+        echo "No machine '$machine' in this test."
+        exit 1
+      fi
+
+      if ! ssh -F ${sshConfig} $machine true; then
+        echo "Couldn't connect to '$machine'. Make sure you've started it with "'`'"$machine.start()"'`'" in the test interactive driver."
+        echo "Problems could also arise if you have modified openssh settings or set a root password."
+        exit 1
+      fi
+
+      # taken from nixos-rebuild (we only want to do the activate part)
+       cmd=(
+          "systemd-run"
+          "-E" "LOCALE_ARCHIVE"
+          "--collect"
+          "--no-ask-password"
+          "--pty"
+          "--quiet"
+          "--same-dir"
+          "--service-type=exec"
+          "--unit=nixos-rebuild-switch-to-configuration"
+          "--wait"
+          "''${configPaths[$machine]}/bin/switch-to-configuration"
+          "test"
+      )
```

Comment:
- Why `test` instead of `switch`? Shouldn't a reboot of the VM boot the new config?
- This `systemd-run` invocation should not be duplicated, but be factored out. I've proposed `${toplevel}/bin/apply` in https://github.com/NixOS/nixpkgs/issues/266290

Reply:
- `test` rather than `switch` because the VMs don't have bootloaders, so `switch` will fail.
- are you proposing we hold off on this until #266290 is merged, or that I should work on it myself, or that there's some other way we should factor it out in the meantime in this PR? I'm happy with any of them (except holding off indefinitely if it's not likely to be merged soon)